File: exercise/19/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head) {
            return head;
        }
        ListNode **pri = &head, **now = &head;
        for (int i = 0; i < n; i++) {
            pri = &((*pri)->next);
        }
        while (*pri) {
            now = &(*now)->next;
            pri = &(*pri)->next;
        }
        ListNode *temp = *now;
        *now = (*now)->next;
        delete temp;
        return head;
    }
};

File: exercise/19/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        if not head:
            return None
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        for _ in range(n + 1):
            first = first.next
        second = dummy
        while first:
            second = second.next
            first = first.next
        second.next = second.next.next
        return dummy.next
